### PR TITLE
support online & ofline distributed  inference

### LINKED ISSUE #142: [New Feature] support distributed inference offline and online (eval while train)
**If this is your first time, please read our contributor guidelines:** https://gitee.com/mindspore/mindspore/blob/master/CONTRIBUTING.md

**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by Mark-ZhouWX**
Thank you for your contribution to the MindYOLO repo.
Before submitting this PR, please make sure:

- [x] You have read the [Contributing Guidelines on pull requests](https://github.com/mindspore-lab/mindyolo/blob/master/CONTRIBUTING.md)
- [x] Your code builds clean without any errors or warnings
- [x] You are using approved terminology
- [ ] You have added unit tests

## Motivation
1. support online & ofline distributed  inference to increase inference speed
(Write your motivation for proposed changes here.)

## Test Plan

(How should this PR be tested? Do you require special setup to run the test or repro the fixed bug?)

## Related Issues and PRs
closes #142 
(Is this PR part of a group of changes? Link the other relevant PRs and Issues here. Use https://help.github.com/en/articles/closing-issues-using-keywords for help on GitHub syntax)

### CODE REVIEW

File: mindyolo/utils/logger.py [hunk 1]
```diff
@@ -37,6 +37,19 @@ def __init__(self, logger_name="MindYOLO"):
         self.device_per_servers = 8
         self.formatter = logging.Formatter("%(asctime)s [%(levelname)s] %(message)s")
 
+    def write(self, msg):
```
Comment (CaitinZhao) [inline mindyolo/utils/logger.py:40]: 这个是啥方法?和info有啥不一样？

File: mindyolo/utils/utils.py [hunk 2]
```diff
@@ -31,7 +31,8 @@ def set_default(args):
     if args.is_parallel:
         init()
         args.rank, args.rank_size, parallel_mode = get_rank(), get_group_size(), ParallelMode.DATA_PARALLEL
-        context.set_auto_parallel_context(device_num=args.rank_size, parallel_mode=parallel_mode, gradients_mean=True)
+        context.set_auto_parallel_context(device_num=args.rank_size, parallel_mode=parallel_mode, gradients_mean=True,
+                                          parameter_broadcast=True)
```
Comment (CaitinZhao) [inline mindyolo/utils/utils.py:35]: 为什么要加这个，cv那边加了这个后精度掉了

File: mindyolo/utils/utils.py [hunk 3]
```diff
@@ -55,7 +56,9 @@ def set_default(args):
         args.config,
     )
     # Directories and Save run settings
-    args.save_dir = os.path.join(args.save_dir, datetime.now().strftime("%Y.%m.%d-%H_%M_%S"))
+    time = get_broadcast_datetime(rank_size=args.rank_size)
```
Comment (CaitinZhao) [inline mindyolo/utils/utils.py:59]: 需要加判断，单卡用Broadcast会报错

File: test.py [hunk 4]
```diff
@@ -79,7 +90,9 @@ def set_default_test(args):
         args.config,
     )
     # Directories and Save run settings
-    args.save_dir = os.path.join(args.save_dir, datetime.now().strftime("%Y.%m.%d-%H:%M:%S"))
+    time = get_broadcast_datetime(rank_size=args.rank_size)
```
Comment (CaitinZhao) [inline test.py:93]: 同上

File: mindyolo/utils/logger.py (hunk 1 above)
Comment (Mark-ZhouWX) [reply mindyolo/utils/logger.py:40]: 这两个方法（write , flush）可以将三方库（比如coco api）的print信息重定向到mindyolo的logger系统

File: mindyolo/utils/utils.py (hunk 2 above)
Comment (Mark-ZhouWX) [reply mindyolo/utils/utils.py:35]: 我想做一个强制保证。在yolox-tiny上训练了100个epoch测试过，加不加对精度没有影响。 我删掉吧

File: mindyolo/utils/utils.py (hunk 3 above)
Comment (Mark-ZhouWX) [reply mindyolo/utils/utils.py:59]: 做了判断，单卡不会走broadcast

File: test.py (hunk 4 above)
Comment (Mark-ZhouWX) [reply test.py:93]: dito

File: mindyolo/utils/utils.py [hunk 5]
```diff
@@ -143,3 +146,45 @@ def draw_result(img_path, result_dict, data_names, is_coco_dataset=True, save_pa
...
+        self.all_reduce = ops.AllReduce(op=ops.ReduceOp.SUM)
+
+    def construct(self, x):
+        return self.all_reduce(x)
+
+
+class Synchronizer:
+    def __init__(self, rank_size=1):
+        # this init method should be run only once
+        self.all_reduce = AllReduce()
```
Comment (zhanghuiyao) [inline mindyolo/utils/utils.py:179]: 单卡的时候初始化AllReduce算子可能存在问题，可以把rank_sink的判断放到上面

Comment (Mark-ZhouWX) [reply mindyolo/utils/utils.py:179]: 单卡不会走allreduce逻辑；此外，单卡和多卡情况我测试过多次，程序运行正常

Comment (zhanghuiyao) [reply mindyolo/utils/utils.py:178]: > 

这个地方初始化后就把ranksize确定了，可以直接在init做判断，单多/卡都能调用，不用在外面做判断

**Review by zhanghuiyao — APPROVED** [review]

**Review by CaitinZhao — APPROVED** [review]

**Pull request merged**